File: core/src/main/java/ru/pocketbyte/locolaser/resource/entity/ResItem.java
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.resource.entity;

import ru.pocketbyte.locolaser.utils.LogUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Single resource item that contain key, value and comment.
 *
 * @author Denis Shurygin
 */
public class ResItem {

    /**
     * Resource key.
     */
    public final String key;

    
    private final List<ResValue> mutableValues = new ArrayList<>(1);

    /**
     * Resource values.
     */
    public final List<ResValue> values = Collections.unmodifiableList(mutableValues);


    public ResItem(ResItem item) {
        this(item.key);
        mutableValues.addAll(item.mutableValues);
    }

    public ResItem(String key) {
        if (key == null)
            throw new IllegalArgumentException("Key must be not null.");

        this.key = key;
    }

    //TODO docs
    public boolean addValue(ResValue value) {
        boolean isHasNoError = true;

        ResValue oldValue = valueForQuantity(value.quantity);
        if (oldValue != null) {
            isHasNoError = false;
            LogUtils.err("Duplicate quantity! Key= " + this.key + " Quantity=" + value.quantity.toString());
            mutableValues.remove(oldValue);
        }
        mutableValues.add(value);
        return isHasNoError;
    }

    //TODO docs
    public boolean removeValue(ResValue value) {
        return mutableValues.remove(value);
    }

    //TODO docs
    public ResValue removeValueForQuantity(Quantity quantity) {
        for (int i = 0; i < mutableValues.size(); i++) {
            ResValue resValue = mutableValues.get(i);
            if (resValue.quantity.equals(quantity)) {
                return mutableValues.remove(i);
            }
        }
        return null;
    }

    //TODO docs
    public ResValue valueForQuantity(Quantity quantity) {
        for (ResValue resValue: mutableValues) {
            if (resValue.quantity.equals(quantity)) {
                return resValue;
            }
        }
        return null;
    }

    //TODO docs
    public boolean isHasQuantities() {
        return mutableValues.size() > 1 || (mutableValues.size() == 1 && mutableValues.get(0).quantity != Quantity.OTHER);
    }

    //TODO docs
    public ResItem merge(ResItem item) {
        if (item != null) {
            for (ResValue value2 : item.mutableValues) {

                ResValue value1 = this.valueForQuantity(value2.quantity);
                if (value1 != null) {
                    this.removeValue(value1);
                }
                this.addValue(value2);
            }
        }
        return this;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof ResItem) {
            ResItem item = (ResItem) obj;
            if (item.key.equals(key) && item.mutableValues.size() == mutableValues.size()) {
                for (ResValue resValue: item.values) {
                    if(!resValue.equals(valueForQuantity(resValue.quantity)))
                        return false;
                }
                return true;
            }
        }
        return super.equals(obj);
    }
}

File: core/src/main/java/ru/pocketbyte/locolaser/resource/entity/ResItem.kt
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.resource.entity

import ru.pocketbyte.locolaser.entity.Quantity
import ru.pocketbyte.locolaser.utils.LogUtils

import java.util.ArrayList
import java.util.Collections

/**
 * Single resource item that contains a key and a list of values (one per plural quantity).
 *
 * @author Denis Shurygin
 */
class ResItem(
    /** Resource key. */
    val key: String
) {

    private val mutableValues = ArrayList<ResValue>(1)

    /** The list of [ResValue] instances for this item, one per plural quantity. */
    val values: List<ResValue> = Collections.unmodifiableList(mutableValues)


    constructor(item: ResItem) : this(item.key) {
        mutableValues.addAll(item.mutableValues)
    }

    /**
     * Adds [value] to this item. If a value with the same quantity already exists, it is replaced.
     * @param value The value to add.
     * @return true if no duplicate was found, false if an existing value was replaced.
     */
    fun addValue(value: ResValue): Boolean {
        var isHasNoError = true

        val oldValue = valueForQuantity(value.quantity)
        if (oldValue != null) {
            isHasNoError = false
            LogUtils.err("Duplicate quantity! Key= ${this.key} Quantity=${value.quantity}")
            mutableValues.remove(oldValue)
        }
        mutableValues.add(value)
        return isHasNoError
    }

    /**
     * Removes [value] from this item.
     * @return true if the value was present and removed, false otherwise.
     */
    fun removeValue(value: ResValue): Boolean {
        return mutableValues.remove(value)
    }

    /**
     * Removes and returns the [ResValue] with the given [quantity], or null if not found.
     */
    fun removeValueForQuantity(quantity: Quantity): ResValue? {
        for (i in mutableValues.indices) {
            val resValue = mutableValues[i]
            if (resValue.quantity == quantity) {
                return mutableValues.removeAt(i)
            }
        }
        return null
    }

    /**
     * Returns the [ResValue] with the given [quantity], or null if not found.
     */
    fun valueForQuantity(quantity: Quantity): ResValue? {
        for (resValue in mutableValues) {
            if (resValue.quantity == quantity) {
                return resValue
            }
        }
        return null
    }

    /**
     * True if this item has multiple values or its single value has a quantity other than [Quantity.OTHER].
     */
    val isHasQuantities: Boolean
        get() = mutableValues.size > 1 || mutableValues.size == 1 && mutableValues[0].quantity !== Quantity.OTHER

    /**
     * Merges the values from [item] into this item, replacing existing values for the same quantity.
     * @return This item after merging.
     */
    fun merge(item: ResItem?): ResItem {
        if (item != null) {
            for (value2 in item.mutableValues) {
                val value1 = this.valueForQuantity(value2.quantity)
                if (value1 != null) {
                    this.removeValue(value1)
                }
                value1.merge(value2)?.let { this.addValue(it) }
            }
        }
        return this
    }

    override fun equals(other: Any?): Boolean {
        if (other is ResItem) {
            if (other.key == key && other.mutableValues.size == mutableValues.size) {
                return other.values.none { it != valueForQuantity(it.quantity) }
            }
        }
        return super.equals(other)
    }

    override fun toString(): String {
        if (isHasQuantities) {
            val valuesString = values.joinToString(",") {
                it.toString()
            }
            return "[$valuesString]"
        }
        return values[0].toString()
    }
}
